File: xiang.li/121.kt
import kotlin.math.max
class Solution {
    fun maxProfit(prices: IntArray): Int {
        if(prices.size == 0){
            return 0
        }
        var dp = IntArray(prices.size)
        dp[0] = 0
        for(i in 1 until prices.size) {
            dp[i] = max(dp[i], dp[i - 1] + prices[i] - prices[i - 1])
            if (dp[i] > dp[0]) {
                dp[0] = dp[i]
            }
        }
        return dp[0]
    }
}

File: hang.liu/121.js
var maxProfit = function(prices) {
  var result = 0;

  var length = prices.length;

  for (var i = 1; i < length; i++) {
    const profit = prices[i] - prices[i - 1];
    prices[i] = Math.min(prices[i], prices[i - 1]);

    if (profit > result) result = profit;
  }

  return result;
};
